lib/scholar/cluster/gmm.ex

```diff
@@ -0,0 +1,407 @@
+                |> Nx.broadcast({num_gaussians, num_features, num_features}),
+              {k = 0, covariances}
+            },
+            k < num_gaussians do
+        covariance_cholesky = Nx.LinAlg.cholesky(covariances[k])
+
+        precision_cholesky =
+          covariance_cholesky
+          |> Nx.LinAlg.triangular_solve(Nx.eye(num_features, type: type), lower: true)
+          |> Nx.transpose()
+
+        {
+          Nx.put_slice(precisions_cholesky, [k, 0, 0], Nx.new_axis(precision_cholesky, 0)),
+          {k + 1, covariances}
+        }
+      end
+
+    precisions_cholesky
+  end
+
+  defnp estimate_log_prob_resp(x, weights, means, precisions_cholesky) do
+    weighted_log_prob = estimate_weighted_log_prob(x, weights, means, precisions_cholesky)
+    log_prob_norm = logsumexp(weighted_log_prob)
+    log_resp = weighted_log_prob - Nx.new_axis(log_prob_norm, 1)
+    {log_prob_norm, log_resp}
+  end
+
+  defnp estimate_weighted_log_prob(x, weights, means, precisions_cholesky) do
+    estimate_log_prob(x, means, precisions_cholesky) + Nx.log(weights)
+  end
+
+  defnp estimate_log_prob(x, means, precisions_cholesky) do
+    type = Nx.type(x)
+    {num_samples, num_features} = Nx.shape(x)
+    {num_gaussians, _} = Nx.shape(means)
+    log_det = compute_log_det_cholesky(precisions_cholesky)
+
+    {log_prob, _} =
+      while {
+              log_prob =
+                Nx.tensor(0.0, type: type)
+                |> Nx.broadcast({num_gaussians, num_samples}),
+              {k = 0, x, means, precisions_cholesky}
+            },
+            k < num_gaussians do
+        mean = means[k]
+        precision_cholesky = precisions_cholesky[k]
+        y = Nx.dot(x, precision_cholesky) - Nx.dot(mean, precision_cholesky)
+        slice = Nx.new_axis(Nx.sum(y * y, axes: [1]), 0)
+
+        {
+          Nx.put_slice(log_prob, [k, 0], slice),
+          {k + 1, x, means, precisions_cholesky}
+        }
+      end
+
+    -0.5 * (num_features * Nx.log(2 * Nx.Constants.pi()) + Nx.transpose(log_prob)) + log_det
+  end
+
+  defnp logsumexp(weighted_log_prob) do
+    max = Nx.reduce_max(weighted_log_prob, axes: [1])
+
+    weighted_log_prob
+    |> Nx.subtract(Nx.new_axis(max, 1))
+    |> Nx.exp()
+    |> Nx.sum(axes: [1])
+    |> Nx.log()
+    |> Nx.add(max)
+  end
+
+  defnp compute_log_det_cholesky(precisions_cholesky) do
+    precisions_cholesky
+    |> Nx.take_diagonal()
+    |> Nx.log()
+    |> Nx.sum(axes: [1])
+  end
+
+  @doc """
+  Makes predictions with the given `model` on inputs `x`.
+
+  ## Return Values
+
+    It returns a tensor with Gaussian assignments for every input point.
+
+  ## Examples
+
+      iex> key = Nx.Random.key(12)
+      iex> x = Nx.tensor([[1, 2], [1, 4], [1, 0], [10, 2], [10, 4], [10, 0]])
+      iex> model = Scholar.Cluster.GaussianMixture.fit(x, num_gaussians: 2, key: key)
+      iex> Scholar.Cluster.GaussianMixture.predict(model, Nx.tensor([[8, 1], [2, 3]]))
+      Nx.tensor(
+        [1, 0]
+      )
```

Comment:
I don't think this will pass on CI

Reply:
Hmmm why exactly? `mix test` also checks the docstrings and runs the code inside.
It's similar to the k-means implementation:
https://github.com/elixir-nx/scholar/blob/main/lib/scholar/cluster/k_means.ex#L297-L299

Reply:
Yup, this is a doctest feature :)

Reply:
TIL :)